### PR TITLE
En 4744 support mobile sdk

### PR DESCRIPTION
**Pull request by guybapx**

### CODE REVIEW

File: src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java [hunk 1]
```diff
@@ -18,19 +24,36 @@
...
-            pageTemplate = fileName + ext;
+        boolean isCaptchaBlock = context.getBlockAction().equals(BlockAction.CAPTCHA);
+        String blockPage = getPage(context, pxConfig, isCaptchaBlock);
+        String action;
+
+        if (context.isMobileToken()) {
+            action = (isCaptchaBlock) ? ACTION_CAPTCHA : ACTION_BLOCK;
+            String base64Page = Base64.encodeToString(blockPage.getBytes(), false);
+            try {
+                blockPage = new ObjectMapper().writeValueAsString(new MobilePageResponse(action, context.getUuid(), context.getAppId(), base64Page));
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java]: missing collector url

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 2]
```diff
@@ -127,21 +129,24 @@ public PXContext(final HttpServletRequest request, final IPProvider ipProvider,
...
-        this.pxCookies = extractPXCookies(cookie);
-        final String pxCaptchaCookie = extractCookieByKey(cookie, Constants.COOKIE_CAPTCHA_KEY);
-        if (pxCaptchaCookie != null) {
-            this.pxCaptcha = pxCaptchaCookie;
+        this.pxCookieOrig = isMobileToken ? ORIGIN_HEADER : ORIGIN_COOKIE;
+
+        //Get cookies
+        final String cookie = request.getHeader(isMobileToken ? MOBILE_SDK_HEADER : ORIGIN_COOKIE);
+        this.pxCookies = isMobileToken ? extractPXMobileCookie(cookie) : extractPXCookies(cookie);
+        //in mobile remove the cookie header - do not send it in request
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/models/PXContext.java]: this is not relevant + its bad practice to change the headers order
need to remove this in my opinion, but we can ask prod if this is necessary

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 3]
```diff
@@ -127,21 +129,24 @@ public PXContext(final HttpServletRequest request, final IPProvider ipProvider,
...
+            logger.debug(PXLogger.LogReason.DEBUG_MOBILE_SDK_DETECTED);
+            this.isMobileToken = true;
         }
 
-        final String cookie = request.getHeader("cookie");
-        this.pxCookies = extractPXCookies(cookie);
-        final String pxCaptchaCookie = extractCookieByKey(cookie, Constants.COOKIE_CAPTCHA_KEY);
-        if (pxCaptchaCookie != null) {
-            this.pxCaptcha = pxCaptchaCookie;
+        this.pxCookieOrig = isMobileToken ? ORIGIN_HEADER : ORIGIN_COOKIE;
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/models/PXContext.java]: default value is cookie

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 4]
```diff
@@ -127,21 +129,24 @@ public PXContext(final HttpServletRequest request, final IPProvider ipProvider,
...
+
+        //Get cookies
+        final String cookie = request.getHeader(isMobileToken ? MOBILE_SDK_HEADER : ORIGIN_COOKIE);
+        this.pxCookies = isMobileToken ? extractPXMobileCookie(cookie) : extractPXCookies(cookie);
+        //in mobile remove the cookie header - do not send it in request
+        if (isMobileToken) {
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/models/PXContext.java]: this is not a good idea.
we dont need to remove any headers

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 5]
```diff
@@ -162,8 +167,27 @@ private void initContext(final HttpServletRequest request, PXConfiguration pxCon
...
+    private Map<String, String> getHeadersFromRequest(HttpServletRequest request) {
+        HashMap<String, String> headers = new HashMap<>();
+        String name;
+        Enumeration headerNames = request.getHeaderNames();
+
+        while (headerNames.hasMoreElements()) {
+            name = (String) headerNames.nextElement();
+
+            //Support case sensitive mobile header
+            if (name.toLowerCase().equals(MOBILE_SDK_HEADER.toLowerCase())) {
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/models/PXContext.java]: just do all headers to lower case, thats what we do in all enforcers

Comment (nitzanpx) [reply src/main/java/com/perimeterx/models/PXContext.java]: line 180, put name.tolowercase

File: src/main/java/com/perimeterx/models/httpmodels/MobilePageResponse.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package com.perimeterx.models.httpmodels;
```
Comment (nitzanpx) [inline src/main/java/com/perimeterx/models/httpmodels/MobilePageResponse.java]: for captcha we're missing collectorUrl

**Comment by nitzanpx** [comment]
except comments, which we'll go over again
need to verify captcha is being solved, block is being displayed and we receive activities properly

File: src/main/java/com/perimeterx/models/PXContext.java (hunk 3 above)
Comment (guybapx) [reply src/main/java/com/perimeterx/models/PXContext.java]: It is
mobile-> "header"
non-mobile -> "cookie" (ORIGIN_COOKIE)
It's better to assign in this way.
In this way we're assign 'pxCookieOrig' value in one place in code.

File: src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java (hunk 1 above)
Comment (guybapx) [reply src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java]: @nitzanpx 
Will it always be 
"https://collector-" + APP_ID + ".perimeterx.net" ?
What about staging and development environments? should we handle only one URL?

File: src/main/java/com/perimeterx/models/httpmodels/MobilePageResponse.java (hunk 6 above)
Comment (guybapx) [reply src/main/java/com/perimeterx/models/httpmodels/MobilePageResponse.java]: Fixed

File: src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java [hunk 7]
```diff
@@ -18,19 +24,36 @@
...
+            responseWrapper.getWriter().print(blockPage);
         } catch (IOException e) {
             throw new PXException(e);
         }
     }
+
+    private String getPage(PXContext context, PXConfiguration pxConfig, boolean isCaptchaBlock) throws PXException {
+        String pageTemplate = BLOCK_FILE_NAME;
+        if (isCaptchaBlock) {
+            String fileName = pxConfig.getCaptchaProvider().name().toLowerCase();
```
Comment (guybapx) [inline src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java]: need to add mobile mustache files

**Pull request closed without merging**